omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java

```diff
@@ -206,24 +225,46 @@ public String post(@RequestParam(value = "username", required = false) String us
 			}
 		}
 
-		//TODO uncomment this to replace the if clause after it
-		if (sessionLocation != null && sessionLocation.hasTag(EmrApiConstants.LOCATION_TAG_SUPPORTS_LOGIN)) {
-			// Set a cookie, so next time someone logs in on this machine, we can default to that same location
-			Cookie cookie = new Cookie(COOKIE_NAME_LAST_SESSION_LOCATION, sessionLocationId.toString());
-			cookie.setHttpOnly(true);
-			pageRequest.getResponse().addCookie(cookie);
+		try {
+			Context.authenticate(username, password);
+			String locationUserPropertyName = administrationService.getGlobalProperty(ReferenceApplicationConstants.LOCATION_USER_PROPERTY_NAME);
+			if (StringUtils.isNotBlank(locationUserPropertyName)) {
+				if (Context.isAuthenticated() && Context.getUserContext().getAuthenticatedUser() != null) {
+					String locationUuid = Context.getUserContext().getAuthenticatedUser().getUserProperty(locationUserPropertyName);
+					if (StringUtils.isNotBlank(locationUuid)) {
+						sessionLocation = locationService.getLocationByUuid(locationUuid);
+						sessionLocationId = sessionLocation.getLocationId();
+					}
+					else {
+						if(sessionLocationId == null ) {
+							pageRequest.getSession().setAttribute(ReferenceApplicationWebConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE,
+									ui.message("referenceapplication.login.error.locationRequired"));
+							// Since the user is already authenticated without location, need to logout before redirecting
+							Context.logout();
+							Map<String, Object> returnParameters = new HashMap<String, Object>();
+							returnParameters.put("showSessionLocations", true);
+							return "redirect:" + ui.pageLink(ReferenceApplicationConstants.MODULE_ID, "login", returnParameters);
+						}
+                    }
+				}
+			}
+
 
-			try {
-				Context.authenticate(username, password);
 
+			//TODO uncomment this to replace the if clause after it
```

Comment:
When do you plan to work on the above TODO?

Reply:
Actually this was came from the exiting code. I just moved that whole part. Do you want me to clear this as well?

Reply:
Just removed @dkayiwa